File: rust/src/patterns/contracts.rs
// #exonware/xwsystem/rust/src/patterns/contracts.rs
//exonware/xwsystem/patterns/contracts.py
//! Company: eXonware.com
//! Author: Eng. Muhammad AlShehri
//! Email: connect@exonware.com
//! Version: 0.1.0.1
//! Generation Date: September 04, 2025
//! 
//! Pattern contracts and interfaces for XWSystem design patterns.


use std::collections::HashMap;

use crate::defs::{AdapterType, AggregateType, ArchitecturalType, BuilderType, ChainHandlerType, CommandType, ConcurrencyType, ContextType, DecoratorType, FacadeType, FactoryType, HandlerType, IteratorType, MediatorType, MementoType, ObserverType, PatternType, PoolType, PrototypeType, ProxyType, RegistryType, SpecificationType, StateType, StrategyType, ValueObjectType, VisitorType};

/// Interface for handlers in the chain of responsibility pattern.
pub trait IHandler {
    /// Handle the request.
    fn handle(&self, request: T) -> T;

    /// Check if this handler can handle the request.
    fn can_handle(&self, request: T) -> bool;

    /// Set the next handler in the chain.
    fn set_next(&self, handler: String) -> String;

}

/// Interface for handler factories.
pub trait IHandlerFactory {
    /// Create a handler of the specified type.
    fn create_handler(&self, handler_type: String) -> T;

    /// Register a handler class.
    fn register_handler(&self, handler_type: String, handler_class: String) -> ();

    /// Unregister a handler class.
    fn unregister_handler(&self, handler_type: String) -> ();

    /// List all registered handler types.
    fn list_handlers(&self) -> Vec<String>;

    /// Check if a handler type is registered.
    fn has_handler(&self, handler_type: String) -> bool;

}

/// Interface for context managers.
pub trait IContextManager {
    /// Check if context is active.
    fn is_active(&self) -> bool;

    /// Get context data.
    fn get_context_data(&self) -> HashMap<String, serde_json::Value>;

}

/// Interface for object pools.
pub trait IObjectPool {
    /// Get an object from the pool.
    fn get(&self, obj_type: String) -> T;

    /// Release an object back to the pool.
    fn release(&self, obj: T) -> ();

    /// Clear objects from the pool.
    fn clear(&self, obj_type: Option<String>) -> ();

    /// Get pool statistics.
    fn get_stats(&self) -> HashMap<String, serde_json::Value>;

    /// Check if pool is empty for a type.
    fn is_empty(&self, obj_type: String) -> bool;

}

/// Interface for registries.
pub trait IRegistry {
    /// Register a value with a key.
    fn register(&self, key: K, value: V) -> ();

    /// Unregister a value by key.
    fn unregister(&self, key: K) -> V;

    /// Get a value by key.
    fn get(&self, key: K) -> Option<V>;

    /// Check if key exists.
    fn has(&self, key: K) -> bool;

    /// List all keys.
    fn list_keys(&self) -> Vec<K>;

    /// List all values.
    fn list_values(&self) -> Vec<V>;

    /// Clear all entries.
    fn clear(&self) -> ();

}

/// Interface for generic registry implementations with metadata support.
///
/// Root cause: Adding generic type parameter for better type safety.
/// Priority #3: Maintainability - Generic types improve code clarity and type checking.
pub trait IGenericRegistry {
    /// Register an item with optional metadata.
    fn register(&self, name: String, item: T, metadata: Option<HashMap<String, serde_json::Value>>) -> bool;

    /// Unregister an item.
    fn unregister(&self, name: String) -> bool;

    /// Get an item by name.
    fn get(&self, name: String) -> Option<T>;

    /// List all registered names.
    fn list_names(&self) -> Vec<String>;

    /// Check if an item exists.
    fn exists(&self, name: String) -> bool;

    /// Clear all registrations.
    fn clear(&self) -> bool;

}

/// Interface for strategies.
pub trait IStrategy {
    /// Execute the strategy.
    fn execute(&self, context: serde_json::Value) -> serde_json::Value;

    /// Check if strategy can handle context.
    fn can_handle(&self, context: serde_json::Value) -> bool;

    /// Get strategy name.
    fn get_name(&self) -> String;

}

/// Interface for observers.
pub trait IObserver {
    /// Update the observer.
    fn update(&self, subject: String, event: serde_json::Value) -> ();

    /// Get observer ID.
    fn get_id(&self) -> String;

}

/// Interface for subjects.
pub trait ISubject {
    /// Attach an observer.
    fn attach(&self, observer: IObserver) -> ();

    /// Detach an observer.
    fn detach(&self, observer: IObserver) -> ();

    /// Notify all observers.
    fn notify(&self, event: serde_json::Value) -> ();

}

/// Interface for commands.
pub trait ICommand {
    /// Execute the command.
    fn execute(&self) -> serde_json::Value;

    /// Undo the command.
    fn undo(&self) -> serde_json::Value;

    /// Check if command can be undone.
    fn can_undo(&self) -> bool;

    /// Get command description.
    fn get_description(&self) -> String;

}

/// Interface for states.
pub trait IState {
    /// Enter the state.
    fn enter(&self, context: serde_json::Value) -> ();

    /// Exit the state.
    fn exit(&self, context: serde_json::Value) -> ();

    /// Handle an event in this state.
    fn handle(&self, context: serde_json::Value, event: serde_json::Value) -> ();

    /// Get state name.
    fn get_name(&self) -> String;

}

/// Interface for builders.
pub trait IBuilder {
    /// Build the object.
    fn build(&self) -> T;

    /// Reset the builder.
    fn reset(&self) -> String;

    /// Check if builder is in valid state.
    fn is_valid(&self) -> bool;

}

/// Interface for prototypes.
pub trait IPrototype {
    /// Clone the object.
    fn clone(&self) -> T;

    /// Create a deep clone.
    fn deep_clone(&self) -> T;

    /// Create a shallow clone.
    fn shallow_clone(&self) -> T;

}

/// Interface for adapters.
pub trait IAdapter {
    /// Adapt source to target.
    fn adapt(&self, source: serde_json::Value) -> serde_json::Value;

    /// Check if source can be adapted.
    fn can_adapt(&self, source: serde_json::Value) -> bool;

    /// Get source type.
    fn get_source_type(&self) -> String;

    /// Get target type.
    fn get_target_type(&self) -> String;

}

/// Interface for decorators.
pub trait IDecorator {
    /// Decorate the target.
    fn decorate(&self, target: T) -> T;

    /// Remove decoration from target.
    fn undecorate(&self, target: T) -> T;

    /// Check if target is decorated.
    fn is_decorated(&self, target: T) -> bool;

}

/// Interface for proxies.
pub trait IProxy {
    /// Get the real object.
    fn get_real_object(&self) -> T;

    /// Check if real object is accessible.
    fn is_accessible(&self) -> bool;

    /// Get proxy type.
    fn get_proxy_type(&self) -> String;

}

/// Interface for facades.
pub trait IFacade {
    /// Execute an operation through the facade.
    fn execute(&self, operation: String) -> serde_json::Value;

    /// Get list of available operations.
    fn get_available_operations(&self) -> Vec<String>;

    /// Check if operation is supported.
    fn is_operation_supported(&self, operation: String) -> bool;

}

/// Interface for dynamic facades.
pub trait IDynamicFacade {
    /// Get list of available formats.
    fn get_available_formats(&self) -> Vec<String>;

    /// Check if format is available.
    fn has_format(&self, format_name: String) -> bool;

    /// Load data using specified format.
    fn load(&self, source: serde_json::Value, format_name: String) -> serde_json::Value;

    /// Save data using specified format.
    fn save(&self, data: serde_json::Value, target: serde_json::Value, format_name: String) -> ();

}

/// Interface for chain handlers.
pub trait IChainHandler {
    /// Handle the request.
    fn handle(&self, request: serde_json::Value) -> serde_json::Value;

    /// Set the next handler.
    fn set_next(&self, handler: String) -> String;

    /// Check if can handle request.
    fn can_handle(&self, request: serde_json::Value) -> bool;

}

/// Interface for mediators.
pub trait IMediator {
    /// Register a colleague.
    fn register_colleague(&self, colleague_id: String, colleague: serde_json::Value) -> ();

    /// Unregister a colleague.
    fn unregister_colleague(&self, colleague_id: String) -> ();

    /// Send a message between colleagues.
    fn send_message(&self, sender_id: String, receiver_id: String, message: serde_json::Value) -> ();

    /// Broadcast a message to all colleagues.
    fn broadcast_message(&self, sender_id: String, message: serde_json::Value) -> ();

}

/// Interface for mementos.
pub trait IMemento {
    /// Get the saved state.
    fn get_state(&self) -> serde_json::Value;

    /// Get creation timestamp.
    fn get_timestamp(&self) -> f64;

    /// Get memento description.
    fn get_description(&self) -> String;

}

/// Interface for originators.
pub trait IOriginator {
    /// Create a memento.
    fn create_memento(&self) -> IMemento;

    /// Restore from memento.
    fn restore_from_memento(&self, memento: IMemento) -> ();

}

/// Interface for visitors.
pub trait IVisitor {
    /// Visit an element.
    fn visit(&self, element: serde_json::Value) -> serde_json::Value;

    /// Check if can visit element.
    fn can_visit(&self, element: serde_json::Value) -> bool;

}

/// Interface for elements that accept visitors.
pub trait IElement {
    /// Accept a visitor.
    fn accept(&self, visitor: IVisitor) -> serde_json::Value;

}

/// Interface for iterators.
pub trait IIterator {
    /// Check if has next item.
    fn has_next(&self) -> bool;

    /// Reset iterator.
    fn reset(&self) -> ();

}

/// Interface for concurrency control.
pub trait IConcurrencyControl {
    /// Acquire the lock.
    fn acquire(&self) -> ();

    /// Release the lock.
    fn release(&self) -> ();

    /// Check if locked.
    fn is_locked(&self) -> bool;

    /// Try to acquire with timeout.
    fn try_acquire(&self, timeout: Option<f64>) -> bool;

}

/// Interface for architectural patterns.
pub trait IArchitecturalPattern {
    /// Initialize the pattern.
    fn initialize(&self) -> ();

    /// Shutdown the pattern.
    fn shutdown(&self) -> ();

    /// Check if initialized.
    fn is_initialized(&self) -> bool;

    /// Get list of components.
    fn get_components(&self) -> Vec<String>;

}

/// Interface for specifications.
pub trait ISpecification {
    /// Check if candidate satisfies specification.
    fn is_satisfied_by(&self, candidate: serde_json::Value) -> bool;

    /// Create AND specification.
    fn and_specification(&self, other: String) -> String;

    /// Create OR specification.
    fn or_specification(&self, other: String) -> String;

    /// Create NOT specification.
    fn not_specification(&self) -> String;

}

/// Interface for value objects.
pub trait IValueObject {
    /// Check if equal to other.
    fn equals(&self, other: serde_json::Value) -> bool;

    /// Get hash code.
    fn get_hash(&self) -> i64;

    /// Convert to string.
    fn to_string(&self) -> String;

}

/// Interface for aggregates in domain-driven design.
pub trait IAggregate {
    /// Get the aggregate ID.
    fn get_id(&self) -> String;

    /// Get the aggregate version.
    fn get_version(&self) -> i64;

    /// Get uncommitted events.
    fn get_uncommitted_events(&self) -> Vec<serde_json::Value>;

    /// Mark events as committed.
    fn mark_events_as_committed(&self) -> ();

}

/// Interface for design patterns.
pub trait IPattern {
    /// Get pattern type.
    fn get_pattern_type(&self) -> PatternType;

    /// Get pattern name.
    fn get_name(&self) -> String;

    /// Get pattern description.
    fn get_description(&self) -> String;

    /// Check if pattern is applicable to context.
    fn is_applicable(&self, context: serde_json::Value) -> bool;

    /// Apply the pattern to context.
    fn apply(&self, context: serde_json::Value) -> serde_json::Value;

    /// Validate data for pattern application.
    fn validate(&self, data: serde_json::Value) -> bool;

}

File: src/exonware/xwsystem/patterns/contracts.py
#exonware/xwsystem/src/exonware/xwsystem/patterns/contracts.py
#exonware/xwsystem/patterns/contracts.py
"""
Company: eXonware.com
Author: eXonware Backend Team
Email: connect@exonware.com
Version: 0.9.0.40
Generation Date: September 04, 2025
Pattern contracts and interfaces for XWSystem design patterns.
"""

from __future__ import annotations
from typing import Any, Protocol, runtime_checkable

from collections.abc import Callable
# Root cause: Migrating to Python 3.12 built-in generic syntax for consistency
# Priority #3: Maintainability - Modern type annotations improve code clarity
# Import enums from types module
from .defs import (
    PatternType,
    HandlerType,
    ContextType,
    FactoryType,
    PoolType,
    RegistryType,
    StrategyType,
    ObserverType,
    CommandType,
    StateType,
    BuilderType,
    PrototypeType,
    AdapterType,
    DecoratorType,
    ProxyType,
    FacadeType,
    ChainHandlerType,
    MediatorType,
    MementoType,
    VisitorType,
    IteratorType,
    ConcurrencyType,
    ArchitecturalType,
    SpecificationType,
    ValueObjectType,
    AggregateType
)
# ============================================================================
# CORE INTERFACES
# ============================================================================
@runtime_checkable

class IHandler[T](Protocol):
    """Interface for handlers in the chain of responsibility pattern."""

    def handle(self, request: T) -> T:
        """Handle the request."""
        ...

    def can_handle(self, request: T) -> bool:
        """Check if this handler can handle the request."""
        ...

    def set_next(self, handler: IHandler[T]) -> IHandler[T]:
        """Set the next handler in the chain."""
        ...
@runtime_checkable

class IHandlerFactory[T](Protocol):
    """Interface for handler factories."""

    def create_handler(self, handler_type: str, **kwargs) -> T:
        """Create a handler of the specified type."""
        ...

    def register_handler(self, handler_type: str, handler_class: type[T]) -> None:
        """Register a handler class."""
        ...

    def unregister_handler(self, handler_type: str) -> None:
        """Unregister a handler class."""
        ...

    def list_handlers(self) -> list[str]:
        """List all registered handler types."""
        ...

    def has_handler(self, handler_type: str) -> bool:
        """Check if a handler type is registered."""
        ...
@runtime_checkable

class IContextManager(Protocol):
    """Interface for context managers."""

    def __enter__(self) -> IContextManager:
        """Enter the context."""
        ...

    def __exit__(self, exc_type: type[BaseException] | None, 
                exc_val: BaseException | None, 
                exc_tb: Any | None) -> bool:
        """Exit the context."""
        ...

    def is_active(self) -> bool:
        """Check if context is active."""
        ...

    def get_context_data(self) -> dict[str, Any]:
        """Get context data."""
        ...
@runtime_checkable

class IObjectPool[T](Protocol):
    """Interface for object pools."""

    def get(self, obj_type: type[T], *args, **kwargs) -> T:
        """Get an object from the pool."""
        ...

    def release(self, obj: T) -> None:
        """Release an object back to the pool."""
        ...

    def clear(self, obj_type: type[T] | None = None) -> None:
        """Clear objects from the pool."""
        ...

    def get_stats(self) -> dict[str, Any]:
        """Get pool statistics."""
        ...

    def is_empty(self, obj_type: type[T]) -> bool:
        """Check if pool is empty for a type."""
        ...
@runtime_checkable

class IRegistry[K, V](Protocol):
    """Interface for registries."""

    def register(self, key: K, value: V) -> None:
        """Register a value with a key."""
        ...

    def unregister(self, key: K) -> V:
        """Unregister a value by key."""
        ...

    def get(self, key: K) -> V | None:
        """Get a value by key."""
        ...

    def has(self, key: K) -> bool:
        """Check if key exists."""
        ...

    def list_keys(self) -> list[K]:
        """List all keys."""
        ...

    def list_values(self) -> list[V]:
        """List all values."""
        ...

    def clear(self) -> None:
        """Clear all entries."""
        ...
@runtime_checkable

class IGenericRegistry[T](Protocol):
    """
    Interface for generic registry implementations with metadata support.
    Root cause: Adding generic type parameter for better type safety.
    Priority #3: Maintainability - Generic types improve code clarity and type checking.
    """

    def register(self, name: str, item: T, metadata: dict[str, Any] | None = None) -> bool:
        """Register an item with optional metadata."""
        ...

    def unregister(self, name: str) -> bool:
        """Unregister an item."""
        ...

    def get(self, name: str) -> T | None:
        """Get an item by name."""
        ...

    def list_names(self) -> list[str]:
        """List all registered names."""
        ...

    def exists(self, name: str) -> bool:
        """Check if an item exists."""
        ...

    def clear(self) -> bool:
        """Clear all registrations."""
        ...
@runtime_checkable

class IStrategy(Protocol):
    """Interface for strategies."""

    def execute(self, context: Any) -> Any:
        """Execute the strategy."""
        ...

    def can_handle(self, context: Any) -> bool:
        """Check if strategy can handle context."""
        ...

    def get_name(self) -> str:
        """Get strategy name."""
        ...
@runtime_checkable

class IObserver(Protocol):
    """Interface for observers."""

    def update(self, subject: ISubject, event: Any) -> None:
        """Update the observer."""
        ...

    def get_id(self) -> str:
        """Get observer ID."""
        ...
@runtime_checkable

class ISubject(Protocol):
    """Interface for subjects."""

    def attach(self, observer: IObserver) -> None:
        """Attach an observer."""
        ...

    def detach(self, observer: IObserver) -> None:
        """Detach an observer."""
        ...

    def notify(self, event: Any) -> None:
        """Notify all observers."""
        ...
@runtime_checkable

class ICommand(Protocol):
    """Interface for commands."""

    def execute(self) -> Any:
        """Execute the command."""
        ...

    def undo(self) -> Any:
        """Undo the command."""
        ...

    def can_undo(self) -> bool:
        """Check if command can be undone."""
        ...

    def get_description(self) -> str:
        """Get command description."""
        ...
@runtime_checkable

class IState(Protocol):
    """Interface for states."""

    def enter(self, context: Any) -> None:
        """Enter the state."""
        ...

    def exit(self, context: Any) -> None:
        """Exit the state."""
        ...

    def handle(self, context: Any, event: Any) -> None:
        """Handle an event in this state."""
        ...

    def get_name(self) -> str:
        """Get state name."""
        ...
@runtime_checkable

class IBuilder[T](Protocol):
    """Interface for builders."""

    def build(self) -> T:
        """Build the object."""
        ...

    def reset(self) -> IBuilder[T]:
        """Reset the builder."""
        ...

    def is_valid(self) -> bool:
        """Check if builder is in valid state."""
        ...
@runtime_checkable

class IPrototype[T](Protocol):
    """Interface for prototypes."""

    def clone(self) -> T:
        """Clone the object."""
        ...

    def deep_clone(self) -> T:
        """Create a deep clone."""
        ...

    def shallow_clone(self) -> T:
        """Create a shallow clone."""
        ...
@runtime_checkable

class IAdapter(Protocol):
    """Interface for adapters."""

    def adapt(self, source: Any) -> Any:
        """Adapt source to target."""
        ...

    def can_adapt(self, source: Any) -> bool:
        """Check if source can be adapted."""
        ...

    def get_source_type(self) -> type:
        """Get source type."""
        ...

    def get_target_type(self) -> type:
        """Get target type."""
        ...
@runtime_checkable

class IDecorator[T](Protocol):
    """Interface for decorators."""

    def decorate(self, target: T) -> T:
        """Decorate the target."""
        ...

    def undecorate(self, target: T) -> T:
        """Remove decoration from target."""
        ...

    def is_decorated(self, target: T) -> bool:
        """Check if target is decorated."""
        ...
@runtime_checkable

class IProxy[T](Protocol):
    """Interface for proxies."""

    def get_real_object(self) -> T:
        """Get the real object."""
        ...

    def is_accessible(self) -> bool:
        """Check if real object is accessible."""
        ...

    def get_proxy_type(self) -> str:
        """Get proxy type."""
        ...
@runtime_checkable

class IFacade(Protocol):
    """Interface for facades."""

    def execute(self, operation: str, *args, **kwargs) -> Any:
        """Execute an operation through the facade."""
        ...

    def get_available_operations(self) -> list[str]:
        """Get list of available operations."""
        ...

    def is_operation_supported(self, operation: str) -> bool:
        """Check if operation is supported."""
        ...
@runtime_checkable

class IDynamicFacade(Protocol):
    """Interface for dynamic facades."""

    def get_available_formats(self) -> list[str]:
        """Get list of available formats."""
        ...

    def has_format(self, format_name: str) -> bool:
        """Check if format is available."""
        ...

    def load(self, source: Any, format_name: str, **kwargs) -> Any:
        """Load data using specified format."""
        ...

    def save(self, data: Any, target: Any, format_name: str, **kwargs) -> None:
        """Save data using specified format."""
        ...
@runtime_checkable

class IChainHandler(Protocol):
    """Interface for chain handlers."""

    def handle(self, request: Any) -> Any:
        """Handle the request."""
        ...

    def set_next(self, handler: IChainHandler) -> IChainHandler:
        """Set the next handler."""
        ...

    def can_handle(self, request: Any) -> bool:
        """Check if can handle request."""
        ...
@runtime_checkable

class IMediator(Protocol):
    """Interface for mediators."""

    def register_colleague(self, colleague_id: str, colleague: Any) -> None:
        """Register a colleague."""
        ...

    def unregister_colleague(self, colleague_id: str) -> None:
        """Unregister a colleague."""
        ...

    def send_message(self, sender_id: str, receiver_id: str, message: Any) -> None:
        """Send a message between colleagues."""
        ...

    def broadcast_message(self, sender_id: str, message: Any) -> None:
        """Broadcast a message to all colleagues."""
        ...
@runtime_checkable

class IMemento(Protocol):
    """Interface for mementos."""

    def get_state(self) -> Any:
        """Get the saved state."""
        ...

    def get_timestamp(self) -> float:
        """Get creation timestamp."""
        ...

    def get_description(self) -> str:
        """Get memento description."""
        ...
@runtime_checkable

class IOriginator(Protocol):
    """Interface for originators."""

    def create_memento(self) -> IMemento:
        """Create a memento."""
        ...

    def restore_from_memento(self, memento: IMemento) -> None:
        """Restore from memento."""
        ...
@runtime_checkable

class IVisitor(Protocol):
    """Interface for visitors."""

    def visit(self, element: Any) -> Any:
        """Visit an element."""
        ...

    def can_visit(self, element: Any) -> bool:
        """Check if can visit element."""
        ...
@runtime_checkable

class IElement(Protocol):
    """Interface for elements that accept visitors."""

    def accept(self, visitor: IVisitor) -> Any:
        """Accept a visitor."""
        ...
@runtime_checkable

class IIterator[T](Protocol):
    """Interface for iterators."""

    def __next__(self) -> T:
        """Get next item."""
        ...

    def __iter__(self) -> IIterator[T]:
        """Get iterator."""
        ...

    def has_next(self) -> bool:
        """Check if has next item."""
        ...

    def reset(self) -> None:
        """Reset iterator."""
        ...
@runtime_checkable

class IConcurrencyControl(Protocol):
    """Interface for concurrency control."""

    def acquire(self) -> None:
        """Acquire the lock."""
        ...

    def release(self) -> None:
        """Release the lock."""
        ...

    def is_locked(self) -> bool:
        """Check if locked."""
        ...

    def try_acquire(self, timeout: float | None = None) -> bool:
        """Try to acquire with timeout."""
        ...
@runtime_checkable

class IArchitecturalPattern(Protocol):
    """Interface for architectural patterns."""

    def initialize(self) -> None:
        """Initialize the pattern."""
        ...

    def shutdown(self) -> None:
        """Shutdown the pattern."""
        ...

    def is_initialized(self) -> bool:
        """Check if initialized."""
        ...

    def get_components(self) -> list[str]:
        """Get list of components."""
        ...
@runtime_checkable

class ISpecification(Protocol):
    """Interface for specifications."""

    def is_satisfied_by(self, candidate: Any) -> bool:
        """Check if candidate satisfies specification."""
        ...

    def and_specification(self, other: ISpecification) -> ISpecification:
        """Create AND specification."""
        ...

    def or_specification(self, other: ISpecification) -> ISpecification:
        """Create OR specification."""
        ...

    def not_specification(self) -> ISpecification:
        """Create NOT specification."""
        ...
@runtime_checkable

class IValueObject(Protocol):
    """Interface for value objects."""

    def equals(self, other: Any) -> bool:
        """Check if equal to other."""
        ...

    def get_hash(self) -> int:
        """Get hash code."""
        ...

    def to_string(self) -> str:
        """Convert to string."""
        ...
@runtime_checkable

class IAggregate(Protocol):
    """Interface for aggregates in domain-driven design."""

    def get_id(self) -> str:
        """Get the aggregate ID."""
        ...

    def get_version(self) -> int:
        """Get the aggregate version."""
        ...

    def get_uncommitted_events(self) -> list[Any]:
        """Get uncommitted events."""
        ...

    def mark_events_as_committed(self) -> None:
        """Mark events as committed."""
        ...
@runtime_checkable

class IPattern(Protocol):
    """Interface for design patterns."""

    def get_pattern_type(self) -> PatternType:
        """Get pattern type."""
        ...

    def get_name(self) -> str:
        """Get pattern name."""
        ...

    def get_description(self) -> str:
        """Get pattern description."""
        ...

    def is_applicable(self, context: Any) -> bool:
        """Check if pattern is applicable to context."""
        ...

    def apply(self, context: Any) -> Any:
        """Apply the pattern to context."""
        ...

    def validate(self, data: Any) -> bool:
        """Validate data for pattern application."""
        ...
